qtt/tests/test_zi_hdawg8.py

```diff
@@ -0,0 +1,88 @@
+import unittest
+from unittest.mock import MagicMock, call
+
+from qtt.instrument_drivers.virtualAwg.awgs.ZurichInstrumentsHDAWG8 import ZurichInstrumentsHDAWG8
+from qtt.instrument_drivers.virtualAwg.awgs.common import AwgCommonError
+
+
+class TestZurichInstrumentsHDAWG8(unittest.TestCase):
+    def setUp(self):
+        class ZIHDAWG8(MagicMock):
+            pass
+
+        self.awg = ZIHDAWG8()
+        self.zi_hdawg8 = ZurichInstrumentsHDAWG8(self.awg, 0)
+
+    def test_enable_outputs(self):
+        self.zi_hdawg8.enable_outputs()
+        calls = [call.enable_channel(ch) for ch in range(0, 8)]
+        self.awg.assert_has_calls(calls)
+
+        with self.assertRaises(AwgCommonError):
+            self.zi_hdawg8.enable_outputs([0, 1, 2, 3, 8])
+
+        self.zi_hdawg8.enable_outputs([6, 7])
+        calls + [call.enable_channel(ch) for ch in range(6, 7)]
+        self.awg.assert_has_calls(calls)
+
+    def test_disable_outputs(self):
+        self.zi_hdawg8.disable_outputs()
+        calls = [call.disable_channel(ch) for ch in range(0, 8)]
+        self.awg.assert_has_calls(calls)
+
+        with self.assertRaises(AwgCommonError):
+            self.zi_hdawg8.disable_outputs([0, 1, 2, 3, 8])
+
+        self.zi_hdawg8.disable_outputs([6, 7])
+        calls + [call.disable_channel(ch) for ch in range(6, 7)]
+        self.awg.assert_has_calls(calls)
+
+    def test_change_setting(self):
+        self.awg.get.return_value = 0
+        self.zi_hdawg8.change_setting('sampling_rate', 2.4e9)
+        self.assertEqual(self.zi_hdawg8.retrieve_setting('sampling_rate'), 2.4e9)
+
+        with self.assertRaises(ValueError):
+            self.zi_hdawg8.change_setting('gain', 0.5)
+
+    def test_update_sampling_rate(self):
+        sample_rates = [2400000000.0, 1200000000.0, 600000000.0, 300000000.0, 150000000.0, 72000000.0, 37500000.0,
+                        18750000.0, 9400000.0, 4500000.0, 2340000.0, 1200.0, 586000.0, 293000.0]
+
+        for sample_rate in sample_rates:
+            self.zi_hdawg8.update_sampling_rate(sample_rate)
+        calls = [call.set('awgs_0_time', i) for i in range(0, 14)]
+        self.awg.assert_has_calls(calls)
+
+        with self.assertRaises(ValueError):
+            self.zi_hdawg8.update_sampling_rate(99)
+
+    def test_retrieve_sampling_rate(self):
+        self.awg.get.return_value = 5
```

Comment:
The relation between 5 and the 72e6 on the line below is not clear from the unit test alone. Better perhaps something like:
```
sampling_rate_index=5
self.awg.get.return_value = sampling_rate_index
expected_sampling_rate=self.sample_rates[sampling_rate_index]
self.assertEqual(expected_sampling_rate, self.zi_hdawg8.retrieve_sampling_rate())
```